src/Model/Aggregate.php

```diff
@@ -0,0 +1,334 @@
+<?php
+
+declare(strict_types=1);
+
+namespace atk4\data\Model;
+
+use atk4\data\Exception;
+use atk4\data\Field;
+use atk4\data\FieldSqlExpression;
+use atk4\data\Model;
+use atk4\data\Reference;
+use atk4\dsql\Query;
+
+/**
+ * Aggregate model allows you to query using "group by" clause on your existing model.
+ * It's quite simple to set up.
+ *
+ * $aggregate = new Aggregate($mymodel);
+ * $aggregate->groupBy(['first','last'], ['salary'=>'sum([])'];
+ *
+ * your resulting model will have 3 fields:
+ *  first, last, salary
+ *
+ * but when querying it will use the original model to calculate the query, then add grouping and aggregates.
+ *
+ * If you wish you can add more fields, which will be passed through:
+ * $aggregate->addField('middle');
+ *
+ * If this field exist in the original model it will be added and you'll get exception otherwise. Finally you are
+ * permitted to add expressions.
+ *
+ * The base model must not be Union model or another Aggregate model, however it's possible to use Aggregate model as nestedModel inside Union model.
+ * Union model implements identical grouping rule on its own.
+ *
+ * You can also pass seed (for example field type) when aggregating:
+ * $aggregate->groupBy(['first','last'], ['salary' => ['sum([])', 'type'=>'money']];
+ */
+class Aggregate extends Model
+{
+    /** @const string */
+    public const HOOK_AFTER_SELECT = self::class . '@afterSelect';
+
+    /**
+     * @deprecated use HOOK_AFTER_SELECT instead - will be removed dec-2020
+     */
+    public const HOOK_AFTER_GROUP_SELECT = self::HOOK_AFTER_SELECT;
+
+    /**
+     * Aggregate model should always be read-only.
+     *
+     * @var bool
+     */
+    public $read_only = true;
+
+    /** @var Model */
+    public $master_model;
+
+    /** @var string */
+    public $id_field;
```

Comment:
why field like this is declared when extending Model?

Reply:
Added phpdoc for it.